pkg/net/libp2p/libp2ptransport.go

```diff
@@ -0,0 +1,264 @@
+
+type Transport struct {
+	host              host.Host
+	ownID             types.NodeID
+	membership        map[types.NodeID]string
+	incomingMessages  chan *events.EventList
+	outboundStreamsMx sync.Mutex
+	outboundStreams   map[types.NodeID]network.Stream
+	logger            logging.Logger
+}
+
+func NewTransport(h host.Host, membership map[types.NodeID]string, ownID types.NodeID, logger logging.Logger) *Transport {
+	if logger == nil {
+		logger = logging.ConsoleErrorLogger
+	}
+	return &Transport{
+		incomingMessages: make(chan *events.EventList),
+		outboundStreams:  make(map[types.NodeID]network.Stream),
+		logger:           logger,
+		membership:       membership,
+		ownID:            ownID,
+		host:             h,
+	}
+}
+
+func (t *Transport) ImplementsModule() {}
+
+func (t *Transport) EventsOut() <-chan *events.EventList {
+	return t.incomingMessages
+}
+
+func (t *Transport) Start() error {
+	t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %s handler starting on %v", t.ownID, t.host.Addrs()))
+	t.host.SetStreamHandler(ID, t.mirHandler)
+	return nil
+}
+
+func (t *Transport) Stop() {
+	t.logger.Log(logging.LevelDebug, "Stopping libp2p transport.")
+	defer t.logger.Log(logging.LevelDebug, "libp2p transport stopped.")
+
+	t.outboundStreamsMx.Lock()
+	defer t.outboundStreamsMx.Unlock()
+
+	for id, s := range t.outboundStreams {
+		if s == nil {
+			continue
+		}
+		t.logger.Log(logging.LevelDebug, "Closing connection", "to", id)
+		if err := s.Close(); err != nil {
+			t.logger.Log(logging.LevelWarn, fmt.Sprintf("Could not close connection to node %v: %v", id, err))
+			continue
+		}
+		t.logger.Log(logging.LevelDebug, "Closed connection", "to", id)
+	}
+
+	if err := t.host.Close(); err != nil {
+		t.logger.Log(logging.LevelError, fmt.Sprintf("Could not close libp2p %v: %v", t.ownID, err))
+	} else {
+		t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %v libp2p host closed", t.ownID))
+	}
+}
+
+func (t *Transport) Connect(ctx context.Context) {
+	wg := sync.WaitGroup{}
+	wg.Add(len(t.membership) - 1)
+
+	for nodeID, nodeAddr := range t.membership {
+		if nodeID == t.ownID {
+			continue
+		}
+
+		go func(nodeID types.NodeID, nodeAddr string) {
+			defer wg.Done()
+
+			t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %s is connecting to node %s", t.ownID.Pb(), nodeID.Pb()))
+
+			maddr, err := multiaddr.NewMultiaddr(nodeAddr)
+			if err != nil {
+				t.logger.Log(logging.LevelError,
+					fmt.Sprintf("failed to parse %v for node %v: %v", nodeAddr, nodeID, err))
+				return
+			}
+
+			// Extract the peer ID from the multiaddr.
+			info, err := peer.AddrInfoFromP2pAddr(maddr)
+			if err != nil {
+				t.logger.Log(logging.LevelError,
+					fmt.Sprintf("failed to parse addr %v for node %v: %v", maddr, nodeID, err))
+				return
+			}
+
+			t.host.Peerstore().AddAddrs(info.ID, info.Addrs, peerstore.PermanentAddrTTL)
+
+			ctx, cancel := context.WithTimeout(ctx, defaultMaxTimeout)
+			defer cancel()
```

Comment:
Is there a purpose for canceling the context here? it seems that when this line is reached, nobody looks at the context anyway any more...